near_deduplication/minhash_deduplication.py

```diff
@@ -0,0 +1,336 @@
+            fn = false_negative_probability(threshold, b, r)
+            error = fp * false_positive_weight + fn * false_negative_weight
+            if error < min_error:
+                min_error = error
+                opt = (b, r)
+    return opt
+
+
+class UnionFind:
+    def __init__(self):
+        self.parent: Dict[int, int] = {}
+
+    def find(self, x):
+        if x not in self.parent:
+            self.parent[x] = x
+        if self.parent[x] != x:
+            self.parent[x] = self.find(self.parent[x])
+        return self.parent[x]
+
+    def union(self, x, y):
+        px = self.find(x)
+        py = self.find(y)
+        self.parent[px] = self.parent[py] = min(px, py)
+
+
+if __name__ == "__main__":
+
+    def run(
+        dataset: str = typer.Option("codeparrot/codeparrot-clean-valid", help="The dataset to use"),  # noqa: E501
+        config: str = typer.Option("default", help="Dataset config"),
+        split: str = typer.Option("train", help="Dataset split"),
+        data_dir: str = typer.Option(None, help="Dataset data directory"),
+        revision: str = typer.Option("main", help="Dataset revision"),
+        column: str = typer.Option("content", help="Dataset column"),
+        cache_dir: str = typer.Option(".cache", help="Cache directory"),
+        ngram_size: int = typer.Option(5, help="The ngram size to use for MinHash"),
+        num_perm: int = typer.Option(256, help="Number of permutations"),
+        threshold: float = typer.Option(0.7, help="Minhash threshold"),
+        output: str = typer.Option(None, help="Store the deduplicated dataset"),
+    ):
+        global uf
+        OUTPUT_BASE = Path(output or "output")
+        OUTPUT_BASE.mkdir(exist_ok=True, parents=True)
+        output = OUTPUT_BASE / "deduplicated"
+
+        logging.basicConfig(level=logging.INFO)
+
+        time_measures = {}
+        start_time = time.time()
+
+        B, R = optimal_param(threshold, num_perm)
+        HASH_RANGES = [(i * R, (i + 1) * R) for i in range(B)]
+        HASH_TABLES = [defaultdict(set) for _ in range(B)]
+
+        time_measures["load_dataset"] = time.time()
+        ds = load_dataset(
+            dataset,
+            config,
+            data_dir=data_dir,
+            split=split,
+            use_auth_token=True,
+            cache_dir=cache_dir,
+            revision=revision,
+            num_proc=os.cpu_count(),
+        )
+        time_measures["load_dataset"] = time.time() - time_measures["load_dataset"]
+        DATA_SIZE = len(ds)
+        PERMUTATIONS = np.array(
+            [
+                (
+                    RNG.randint(1, MERSENNE_PRIME, dtype=np.uint64),
+                    RNG.randint(0, MERSENNE_PRIME, dtype=np.uint64),
+                )
+                for _ in range(num_perm)
+            ],
+            dtype=np.uint64,
+        ).T
+
+        time_measures["minhash"] = time.time()
+        embedded = ds.map(
+            function=embed_func,
```

Comment:
I'm wondering if we can make `embed_func` [batched](https://huggingface.co/docs/datasets/about_map_batch) (takes a list of contents instead of one) to speed things up with `map`

Reply:
I will test it out in another PR.